exonum-java-binding/core/src/main/java/com/exonum/binding/core/runtime/TransactionMethodExtractor.java

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright 2019 The Exonum Team
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.exonum.binding.core.runtime;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.stream.Collectors.toMap;
+
+import com.exonum.binding.core.transaction.TransactionContext;
+import com.exonum.binding.core.transaction.TransactionMethod;
+import com.google.common.annotations.VisibleForTesting;
+
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.MethodHandles.Lookup;
+import java.lang.reflect.Method;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Finds and validates transaction methods in a service.
+ */
+final class TransactionMethodExtractor {
+
+  /**
+   * Returns a map of transaction ids to transaction methods found in a service class.
+   *
+   * @see TransactionMethod
+   */
+  static Map<Integer, MethodHandle> extractTransactionMethods(Class<?> serviceClass) {
+    Map<Integer, Method> transactionMethods = findTransactionMethods(serviceClass);
+    Lookup lookup = MethodHandles.lookup();
+    return transactionMethods.entrySet().stream()
+        .peek(tx -> validateTransactionMethod(tx.getValue(), serviceClass))
+        .collect(toMap(Map.Entry::getKey,
+            (e) -> toMethodHandle(e.getValue(), lookup)));
+  }
+
+  @VisibleForTesting
+  static Map<Integer, Method> findTransactionMethods(Class<?> serviceClass) {
+    Map<Integer, Method> transactionMethods = new HashMap<>();
+    while (serviceClass != Object.class) {
+      Method[] classMethods = serviceClass.getDeclaredMethods();
+      for (Method method : classMethods) {
+        if (method.isAnnotationPresent(TransactionMethod.class)) {
+          TransactionMethod annotation = method.getAnnotation(TransactionMethod.class);
+          int transactionId = annotation.value();
+          checkDuplicates(transactionMethods, transactionId, serviceClass, method);
+          transactionMethods.put(transactionId, method);
+        }
+      }
+      serviceClass = serviceClass.getSuperclass();
+    }
+    return transactionMethods;
+  }
+
+  private static void checkDuplicates(Map<Integer, Method> transactionMethods, int transactionId,
+      Class<?> serviceClass, Method method) {
+    if (transactionMethods.containsKey(transactionId)) {
+      String errorMessage = String.format("Service %s has more than one transaction with the same"
```

Comment:
(Nit): this one is actually first (given the order of traversal). Also, variable.

```suggestion
      String firstMethodName = transactionMethods.get(transactionId).getName();
      String errorMessage = String.format("Service %s has more than one transaction with the same"
              + " id (%s): first: %s; second: %s",
          serviceClass.getName(), transactionId, firstMethodName, method.getName());
```